### PR TITLE
Implement RFQ partial fill

### PR DESCRIPTION
**Pull request by charlesjhongc**
- Based on #209 
- The RFQ should support partial fill but only can be filled once.
- Since the shared `Offer` struct become not enough for different contracts (always need to add more fields and handle the nested struct hash), having separated struct makes more sense now.

### CODE REVIEW

File: contracts/RFQ.sol [hunk 1]
```diff
@@ -51,111 +54,125 @@ contract RFQ is IRFQ, Ownable, TokenCollector, EIP712 {
...
 
     function fillRFQ(
-        Offer calldata offer,
+        RFQOffer calldata rfqOffer,
+        uint256 takerRequestAmount,
         bytes calldata makerSignature,
         bytes calldata makerTokenPermit,
         bytes calldata takerTokenPermit,
         address payable recipient,
         uint256 feeFactor
```
Comment (pilagod) [inline contracts/RFQ.sol:63]: Q: Why not using `RFQTx` here? I see the relevant args are all needed to be passed in too.

**Review by pilagod — COMMENTED** [review]
1. So RFQ could only be filled once, even the partial fill flag is on?
2. It would be better for review by separating refactoring and feature changes. In this PR it is good enough to define RFQ offer without the general one. And a bigger scope refactoring to other contracts could be done in another PR.

File: contracts/RFQ.sol [hunk 2]
```diff
@@ -51,111 +54,125 @@ contract RFQ is IRFQ, Ownable, TokenCollector, EIP712 {
...
+            Address.sendValue(_rfqOffer.maker, _rfqTx.takerRequestAmount);
         } else {
-            _collect(_offer.takerToken, _offer.taker, _offer.maker, _offer.takerTokenAmount, _takerTokenPermit);
+            if (msg.value != 0) revert InvalidMsgValue();
+            _collect(_rfqOffer.takerToken, _rfqOffer.taker, _rfqOffer.maker, _rfqTx.takerRequestAmount, _takerTokenPermit);
         }
 
         // collect makerToken from maker to this
-        _collect(_offer.makerToken, _offer.maker, address(this), _offer.makerTokenAmount, _makerTokenPermit);
+        uint256 makerHelloAmount = _rfqOffer.makerTokenAmount;
```
[the remark this answers is not in the document]
Comment (charlesjhongc) [reply contracts/RFQ.sol:136]: No...forgot to rename it 😱😱. Let's call it `makerSettleAmount`.

**Comment by charlesjhongc** [comment]
Good suggestion by @pilagod . I will open another PR for RFQ changes only and suspend this PR for now.

**Comment by charlesjhongc** [comment]
@pilagod 
Q: So RFQ could only be filled once, even the partial fill flag is on?
A: Yes, any RFQ offer could be filled only once.

**Comment by charlesjhongc** [comment]
Hmmm, not sure why the gas diff is keep failing. Checkout to a new branch and rebase to master and got CI pass #212 . These two branch has no diff. Perhaps we can move to #212 ?

**Pull request closed without merging**